### PR TITLE
feat: yield Action fly-out menu from uischema

### LINKED ISSUE #3455: Allow custom grouping of custom actions in action flyout
We need a way for users to group their custom actions in the flyout menu.

### PR DESCRIPTION
**Pull request by yeze322**
## Description

Closes #3455 

Yield Action fly-out menu from uischema.

#### Sample menu config in uischema:
  ```json
{
    "Microsoft.SendActivity": {
      "label": "Send a response",
      "submenu": false
    },
    "Microsoft.TextInput": {
      "label": "Text",
      "submenu": ["Ask a question"]
    },
    "Microsoft.NumberInput": {
      "label": "Number",
      "submenu": ["Ask a question"]
    }
}
  ```
#### Full menu schema definition
[defaultMenuSchema.ts](https://github.com/microsoft/BotFramework-Composer/blob/ff5a57e25853a1dc65bdda5ed2257bc94202da0e/Composer/packages/ui-plugins/composer/src/defaultMenuSchema.ts)

#### The screenshot of generated menu:
image

#### User story - group a custom action
1. In ejected runtime, add a custom Action and run schema merge tool.
2. Now the custom $kind should appears in the flyout under the `Custom Action` label.
  image
3. To group that $kind, update the `sdk.uischema` file. (under `BOT_NAME/schemas/sdk.uischema`)
  For example, to put it under `"Looping"`, the config is
  ```json
  {
    "menu": {
       "label": "My custom action",
       "submenu": ["Looping"]
    }
  }
  ```
  image
4. Now the custom $kind should appear under the assigned label ("Looping")
  image

### CODE REVIEW

**Comment by yeze322** [comment]
I can think of a solution that providing a `$order` property in 'sdk.uischema' file. It's typed with `string[]`, stores the in-order traversal result of the action menu tree.

For example, to describe our default action menu, the order string is:
```
"$order": [
  // L1 labels
  "Send a response",
  "Ask a question",
  "Create a condition",
  ...,
  // L2 labels (under 'Ask a question')
  "Text",
  "Number",
  ...,
  // L2 labels (under 'Create a condition'),
  "Branch: If/else",
  "Branch: Switch",
  ...
]
```


This `$order` field is saved in sdk.uischema file as a global config.
```
{
  "$schema": "...",
  "Microsoft.SendActivity": {
    "form": {...},
    "menu": {...},
  },
  ...,
  "$menu": {
    // the $order array
    "$order": [
      "Send a response",
      ...,
    ]
   }
}
```

**Comment by a-b-r-o-w-n** [comment]
@yeze322 I like this approach, but it makes me want to just have 1 way to configure the menu. My concern with a top-level configuration is ambiguity around how to merge / modify it. If I create a custom action, how do I customize the menu? Or if I want to re-order 2 items, can I do so without needing to copy all of the config?

**Comment by yeze322** [comment]
@a-b-r-o-w-n That's good questions. Defining a global menu config seems violates the schema merge strategy and hard to maintain though it solves the menu order problem.

I misunderstood that there should be a initial sdk.schema file as a template where we put the $order config, however, I realized it's not how schema merge works - the only way we preserve the $order config is to hardcode it in the schema merge script, that's not a good solution.

This solution cannot solve

**Comment by yeze322** [comment]
Another solution I can imagine is to assign a `order` number to each $kind. As an example:
```json
{
  "Microsoft.SendActivity": {
    "label": "Send a response",
    "submenu": false,
    "order": 1
  },
  "Microsoft.TextInput": {
    "label": "Text",
    "submenu": ["Ask a question"],
    "order": 21
  }
}
```

**Comment by yeze322** [comment]
Finally, the order of menu items is kept in FlowEditor - [defaultMenuOrder.ts](https://github.com/microsoft/BotFramework-Composer/blob/zeye/schema-menu/Composer/packages/extensions/adaptive-flow/src/adaptive-flow-editor/renderers/EdgeMenu/defaultMenuOrder.ts) file as a built-in config.

We can either keep that config as it is or move to uischema file in the future.

**Comment by a-b-r-o-w-n** [comment]
We will continue to discuss options on how to represent the order in a declarative way. Our solution may be a composer only configuration.

File: Composer/packages/client/src/pages/design/PropertyEditor.tsx [hunk 1]
```diff
@@ -57,19 +57,14 @@ const PropertyEditor: React.FC = () => {
     };
   }, [formData]);
 
+  const { plugins: pluginConfig } = useContext(ExtensionContext);
```
[the remark this answers is not in the document]
Comment (yeze322) [reply Composer/packages/client/src/pages/design/PropertyEditor.tsx:60]: updated

File: Composer/packages/client/src/pages/design/PropertyEditor.tsx [hunk 2]
```diff
@@ -3,18 +3,18 @@
...
 import AdaptiveForm, { resolveRef, getUIOptions } from '@bfc/adaptive-form';
-import Extension, { FormErrors, JSONSchema7, PluginConfig } from '@bfc/extension';
+import { FormErrors, JSONSchema7 } from '@bfc/extension';
 import formatMessage from 'format-message';
 import isEqual from 'lodash/isEqual';
 import debounce from 'lodash/debounce';
 import mapValues from 'lodash/mapValues';
 import { Resizable, ResizeCallback } from 're-resizable';
 import { MicrosoftAdaptiveDialog } from '@bfc/shared';
+import ExtensionContext from '@bfc/extension/lib/extensionContext';
```
[the remark this answers is not in the document]
Comment (yeze322) [reply Composer/packages/client/src/pages/design/PropertyEditor.tsx:15]: updated

File: Composer/packages/extensions/adaptive-flow/src/adaptive-flow-editor/renderers/EdgeMenu/EdgeMenu.tsx [hunk 3]
```diff
@@ -50,6 +51,10 @@ export const EdgeMenu: React.FC<EdgeMenuProps> = ({ id, onClick }) => {
     setMenuSelected(menuSelected);
   };
 
+  const {
+    plugins: { uiSchema },
+  } = useContext(ExtensionContext);
```
[the remark this answers is not in the document]
Comment (yeze322) [reply Composer/packages/extensions/adaptive-flow/src/adaptive-flow-editor/renderers/EdgeMenu/EdgeMenu.tsx]: Yeah, that's right, I should leverage those shared hooks here.

[the remark this answers is not in the document]
Comment (yeze322) [reply Composer/packages/extensions/adaptive-flow/src/adaptive-flow-editor/renderers/EdgeMenu/EdgeMenu.tsx]: Updated. Applied the `useMenuConfig` hook.

**Comment by yeze322** [comment]
@a-b-r-o-w-n  thanks for the review! I replaced all directly usages of `ExtensionContext` to proper extension hooks.

Now the usages are
- Form Editor uses `useFormConfig` to draw Form UI
- Flow Editor (`EdgeMenu.tsx`) uses `useMenuConfig` to draw Action flyout menu
- `useFlowConfig` will be used in FlowEditor in the future to draw an Action node

**Pull request merged**